### PR TITLE
feat: support configurable number of bits

### PR DESCRIPTION
**Pull request by dickens7**
https://github.com/apache/apisix/issues/4209

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 build
-.idea
\ No newline at end of file
+.idea
+.vscode
```
Comment (Yiyiyimu) [inline .gitignore]: need to add an extra newline here

File: README.md [hunk 2]
```diff
@@ -23,15 +23,19 @@ Usage
...
 
-where
+where default
 
 * `s` is a 12-bit integer that increments when next_id() is called multiple times for the same millisecond
 * `n` is a 5-bit integer representing the node within a given data center
 * `d` is a 5-bit integer representing a unique data center or group of servers
 * `t` is a 42-bit integer representing the current timestamp in milliseconds
-    * the number of milliseconds to have elapsed since 1413817200000 or 2014-10-20T15:00:00.000Z
+    * the number of milliseconds to have elapsed since 1609430400000 or 2021-01-01T00:00:00Z
```
Comment (Yiyiyimu) [inline README.md]: It should be `1609459200000` since `2021-01-01T00:00:00Z` is UTC time
```suggestion
    * the number of milliseconds to have elapsed since 1609459200000 or 2021-01-01T00:00:00Z
```

File: src/main.c [hunk 3]
```diff
@@ -15,8 +15,8 @@ static int g_node_id = 0;
 static long g_last_timestamp = -1;
 static int g_sequence = 0;
 
-// 2014-10-20T15:00:00Z
-#define SNOWFLAKE_EPOC 1413817200000L
+// 2021-01-01T00:00:00Z
+#define SNOWFLAKE_EPOC 1609430400000L
```
Comment (Yiyiyimu) [inline src/main.c]:
```suggestion
#define SNOWFLAKE_EPOC 1609459200000L
```

File: src/main.c [hunk 4]
```diff
@@ -27,59 +27,95 @@ static int g_sequence = 0;
...
+{
     struct timeval tv;
     gettimeofday(&tv, NULL);
 
     return tv.tv_sec * 1000 + tv.tv_usec / 1000;
 }
 
-static long get_til_next_millis(long last_timestamp) {
+static long get_til_next_millis(long last_timestamp)
+{
```
Comment (Yiyiyimu) [inline src/main.c]: Do you mind change your indentation setup temporarily so there would be less unnecessary changes

File: rockspec/api7-snowflake-2.0-1.rockspec [hunk 5]
```diff
@@ -1,16 +1,16 @@
-package = "snowflake"
```
Comment (spacewander) [inline rockspec/api7-snowflake-2.0-1.rockspec:1]: We can leave this file unchanged as now we start from v2.1

File: src/main.c [hunk 6]
```diff
@@ -27,59 +27,95 @@ static int g_sequence = 0;
...
+    {
         g_sequence = 0;
     }
 
     g_last_timestamp = ts;
-    ts = ((ts - SNOWFLAKE_EPOC) << TIMESTAMP_SHIFT)
-            | (g_datacenter_id << DATACENTER_ID_SHIFT)
-            | (g_node_id << NODE_ID_SHIFT)
-            | g_sequence;
+    ts = ((ts - conf.snowflake_epoc) << conf.timestamp_shift) | (g_datacenter_id << conf.datacenter_id_shift) | (g_node_id << conf.node_id_shift) | g_sequence;
```
Comment (spacewander) [inline src/main.c]: Better to break it into multiple lines

File: src/main.c [hunk 7]
```diff
@@ -27,59 +27,95 @@ static int g_sequence = 0;
...
 
-static int luasnowflake_init(lua_State *L) {
+static int luasnowflake_init(lua_State *L)
+{
+    conf.snowflake_epoc = luaL_optlong(L, 3, SNOWFLAKE_EPOC);
+    conf.node_id_bits = luaL_optint(L, 4, NODE_ID_BITS);
+    conf.datacenter_id_bits = luaL_optint(L, 5, DATACENTER_ID_BITS);
+    conf.sequence_bits = luaL_optint(L, 6, SEQUENCE_BITS);
+
+    if ((conf.node_id_bits + conf.datacenter_id_bits + conf.sequence_bits) > 32)
```
Comment (spacewander) [inline src/main.c]: Should we also count the snowflake_epoc?
We should also ensure all fields are >= 1

File: src/main.c [hunk 8]
```diff
@@ -27,59 +27,95 @@ static int g_sequence = 0;
 
 #define SEQUENCE_MASK (0xffffffff ^ (0xffffffff << SEQUENCE_BITS))
```
Comment (spacewander) [inline src/main.c]: This macro is unused?

File: src/main.c [hunk 9]
```diff
@@ -27,59 +27,95 @@ static int g_sequence = 0;
...
+
+    if ((conf.node_id_bits + conf.datacenter_id_bits + conf.sequence_bits) > 32)
+    {
+        return luaL_error(L, "(node_id_bits + datacenter_id_bits + sequence_bits) cannot be greater than 32");
+    }
+
     g_datacenter_id = luaL_checkint(L, 1);
-    if (g_datacenter_id < 0x00 || g_datacenter_id > 0x1f) {
-        return luaL_error(L, "datacenter_id must be an integer n, where 0 ≤ n ≤ 0x1f");
+    if (g_datacenter_id < 0x00 || g_datacenter_id > (1 << conf.datacenter_id_bits))
```
Comment (spacewander) [inline src/main.c]: Look like it should be `1 << (conf.datacenter_id_bits - 1)`?

File: rockspec/api7-snowflake-2.0-1.rockspec (hunk 5 above)
Comment (dickens7) [reply rockspec/api7-snowflake-2.0-1.rockspec:1]: Can I publish to luarocks without changing the version number?

Comment (dickens7) [reply rockspec/api7-snowflake-2.0-1.rockspec:1]: Is it possible to publish to luarocks after only changing the version number?

Comment (spacewander) [reply rockspec/api7-snowflake-2.0-1.rockspec:1]: Yes.

File: src/main.c [hunk 10]
```diff
@@ -44,16 +49,51 @@ static long get_til_next_millis(long last_timestamp) {
...
+    if (conf.datacenter_id_bits < 1) {
+        return luaL_error(L, "datacenter_id_bits must be an integer n where ≥ 1");
+    }
+
+    conf.sequence_bits = luaL_optint(L, 6, SEQUENCE_BITS);
+    if (conf.sequence_bits < 1) {
+        return luaL_error(L, "sequence_bits must be an integer n where ≥ 1");
+    }
+
+    if ((conf.node_id_bits + conf.datacenter_id_bits + conf.sequence_bits) > 32) {
```
Comment (spacewander) [inline src/main.c:77]: Could you explain why we need the timestamp >= 32 bits

File: README.md [hunk 11]
```diff
@@ -23,15 +23,19 @@ Usage
...
 * `t` is a 42-bit integer representing the current timestamp in milliseconds
-    * the number of milliseconds to have elapsed since 1413817200000 or 2014-10-20T15:00:00.000Z
+    * the number of milliseconds to have elapsed since 1609459200000 or 2021-01-01T00:00:00.000Z
 
-`sf.init(datacenter_id, node_id)` is used to initialize snowflake and set values for ddddd and nnnnn as follows:
+`sf.init(datacenter_id, node_id, snowflake_epoc, node_id_bits, datacenter_id_bits, sequence_bits)` is used to initialize snowflake and set values for ddddd and nnnnn as follows:
 
-* `datacenter_id` is an integer n, where 0 ≤ n ≤ 0x1f and specifies the `ddddd` portion of the id
-* `node_id` is an integer n, where 0 ≤ n ≤ 0x1f and specifies the `nnnnn` portion of the id
+* `datacenter_id` is an integer n, where 0 ≤ n ≤ (1 << datacenter_id_bits) and specifies the `ddddd` portion of the id
```
Comment (spacewander) [inline README.md]:
```suggestion
* `datacenter_id` is an integer n, where 0 ≤ n < (1 << datacenter_id_bits) and specifies the `ddddd` portion of the id
```

File: README.md [hunk 12]
```diff
@@ -23,15 +23,19 @@ Usage
...
+* `node_id` is an integer n, where 0 ≤ n ≤ (1 << node_id_bits) and specifies the `nnnnn` portion of the id
```
Comment (spacewander) [inline README.md]:
```suggestion
* `node_id` is an integer n, where 0 ≤ n < (1 << node_id_bits) and specifies the `nnnnn` portion of the id
```

File: src/main.c (hunk 10 above)
Comment (dickens7) [reply src/main.c:77]: The timestamp >= 32 is because it is in milliseconds and only lasts 49 days when it equals 32 bits

Comment (spacewander) [reply src/main.c:77]: I see. Can you add it as a comment?

File: rockspec/snowflake-2.1-0.rockspec [hunk 13]
```diff
@@ -0,0 +1,28 @@
+package = "snowflake"
```
Comment (spacewander) [inline rockspec/snowflake-2.1-0.rockspec]: The new package requires a "api7-" prefix, like this one: https://github.com/api7/lua-resty-http/blob/master/api7-lua-resty-http-0.1.0-0.rockspec

Comment (dickens7) [reply rockspec/snowflake-2.1-0.rockspec]: Does the c library need to add a prefix? If someone uses snowflake.so, there may still be conflicts

File: src/main.c (hunk 10 above)
Comment (dickens7) [reply src/main.c:77]: added

File: rockspec/snowflake-2.1-0.rockspec (hunk 13 above)
Comment (spacewander) [reply rockspec/snowflake-2.1-0.rockspec]: I think it is fine, nobody will use two snowflake libraries.

**Review by spacewander — APPROVED** [review]

File: rockspec/snowflake-2.1-0.rockspec (hunk 13 above)
Comment (dickens7) [reply rockspec/snowflake-2.1-0.rockspec]: I think we can change the default value of `SNOWFLAKE_EPOC` back to the original value, so that it can still be compatible with the original library.
> 1609459200000                      =>    1413817200000
> 2021-01-01T00:00:00.000Z    =>     2014-10-20T15:00:00.000Z

**Review by Yiyiyimu — APPROVED** [review]

**Comment by spacewander** [comment]
@membphis 

Would you spend some time to make a 2.1 release? Thanks!

**Comment by membphis** [comment]
> @membphis
> 
> Would you spend some time to make a 2.1 release? Thanks!

done

image
